pkg/volume/util/quota/common/quota_linux_common_impl.go

```diff
@@ -0,0 +1,185 @@
+	QXGetPQuota = 0x00580302
+	// QXSetPQLim -- Set project quotas for XFS filesystems
+	QXSetPQLim = 0x00580402
+	// XFSProjQuota -- use of project (vs. user or group) quotas
+	XFSProjQuota = 0x00000002
+)
+
+// Shadowing struct fsxattr from <linux.fs.h>
+type fsxattr struct {
+	xflags     uint32
+	extsize    uint32
+	nextents   uint32
+	projid     uint32
+	cowextsize uint32
+	pad        [8]byte
+}
+
+// Shadowing fs_quota_stat_t from <linux/dqblk_xfs.h>
+type fsqfilestat struct {
+	fsino     uint64
+	fsnblks   uint64
+	fsextents uint64
+}
+
+type fsquotastat struct {
+	version      int8
+	flags        uint16
+	pad          int8
+	uquota       fsqfilestat
+	gquota       fsqfilestat
+	incoredqs    uint32
+	btimelimit   int32
+	itimelimit   int32
+	rtbtimelimit int32
+	bwarnlimit   int16
+	iwarnlimit   int16
+}
+
+// IsFilesystemOfType determines whether the filesystem specified is of the type
+// specified by the magic number
+func IsFilesystemOfType(mountpoint string, backingDev string, magic int64) bool {
+
+	var buf syscall.Statfs_t
+	err := syscall.Statfs(mountpoint, &buf)
+	if err != nil {
+		klog.V(3).Infof("Extfs Unable to statfs %s: %v", mountpoint, err)
+		return false
+	}
+	if buf.Type != magic {
+		return false
+	}
+
+	var qstat fsquotastat
+	CPath := append([]byte(backingDev), 0)
+
+	_, _, errno := unix.Syscall6(unix.SYS_QUOTACTL, uintptr(XGetQStatPrjQuota), uintptr(unsafe.Pointer(&CPath[0])), 0, uintptr(unsafe.Pointer(&qstat)), 0, 0)
+	return errno == 0 && qstat.flags&FSQuotaPDQEnfd > 0 && qstat.flags&FSQuotaPDQAcct > 0
+}
+
+func openDir(path string) (*os.File, error) {
+
+	dir, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("Can't open %s: %v", path, err)
+	}
+	s, err := dir.Stat()
+	if err != nil {
+		dir.Close()
+		return nil, fmt.Errorf("Can't stat %s: %v", path, err)
+	}
+	if !s.IsDir() {
+		dir.Close()
+		return nil, fmt.Errorf("%s: not a directory", path)
+	}
+	return dir, nil
+}
+
+func closeDir(dir *os.File) {
+	if dir != nil {
+		dir.Close()
+	}
+}
+
+// GetQuotaOnDir retrieves the quota ID (if any) associated with the specified directory
+func GetQuotaOnDir(path string) (QuotaID, error) {
+	dir, err := openDir(path)
+	if err != nil {
+		klog.V(3).Infof("Can't open directory %s: %#+v", path, err)
+		return BadQuotaID, err
+	}
+	defer closeDir(dir)
+	var fsx fsxattr
+	_, _, errno := unix.Syscall(unix.SYS_IOCTL, dir.Fd(), FSIocGetXAttr,
+		uintptr(unsafe.Pointer(&fsx)))
+	if errno != 0 {
+		return BadQuotaID, fmt.Errorf("Failed to get quota ID for %s: %v", path, errno.Error())
+	}
+	if fsx.projid == 0 {
+		return BadQuotaID, fmt.Errorf("Failed to get quota ID for %s: %s", path, "no applicable quota")
+	}
+	return QuotaID(fsx.projid), nil
+}
+
+// ApplyProjectToDir applies the specified quota ID to the specified directory
+func ApplyProjectToDir(path string, id QuotaID) error {
+	dir, err := openDir(path)
+	if err != nil {
+		return err
+	}
+	defer closeDir(dir)
```

Comment:
closeDir can be removed in favor of: `defer dir.Close()`

Reply:
Yes; I did this quick and dirty as a proof of concept.